### PR TITLE
grub: add grub.useOsprober option

### LINKED ISSUE #7406: grub2: Auto-detect other Linux and Windows installations?
This is a commonly asked for feature in #nixos and I'd be interested in seeing this as well! I'm thinking of something similar to os-prober on Arch Linux.

I'm just not sure where to begin implementing such a thing. Should I attempt to integrate os-prober? Is there some obvious solution I am missing? Any feedback would be appreciated!

### PR DESCRIPTION
**Pull request by symphorien**
###### Motivation for this change

Fix #7406

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I only tested on one machine (BIOS, x86_64 and grub2).

### CODE REVIEW

File: nixos/modules/system/boot/loader/grub/grub.nix [hunk 1]
```diff
@@ -265,6 +267,14 @@ in
         '';
       };
 
+      useOsprober = mkOption {
```
Comment (edolstra) [inline nixos/modules/system/boot/loader/grub/grub.nix]: Maybe rename this to `useOSProber`?

**Comment by edolstra** [comment]
Looks good to me.

**Comment by symphorien** [comment]
@globin : you didn't rename the option in https://github.com/NixOS/nixpkgs/pull/22558/files#diff-3f0e9813c2baa6a4e7f8dbe5fd38ea56R435
Maybe get is case insensitive; I can't test on my machine it would entail a few hours of compilation and I can't reopen the PR anyway.

**Comment by globin** [comment]
I fixed that in an extra commit :) on my phone right now but should be the commit right after the merge

**Comment by vcunat** [comment]
Nix is case-sensitive.

**Comment by florianjacob** [comment]
@symphorien or anybody else, I'm trying to test this, I added `useOSProber = true;` to my grub config, the output of nixos-rebuild seems like it probes for other OSes, but the resulting /boot/grub/grub.cfg is the same as without that option:

```
[florian@nausikaa:/]$ sudo nixos-rebuild switch
building Nix...
building the system configuration...
updating GRUB 2 menu...
lsblk: /dev/mapper/no*[0-9]: not a block-oriented device
lsblk: /dev/mapper/raid*[0-9]: not a block-oriented device
lsblk: /dev/mapper/disks*[0-9]: not a block-oriented device
activating the configuration...
setting up /etc...
```

I have an Arch Linux laying around on that hard disk as well. Is there a possibility to run the prober script alone to see what's its problem, or another way to debug this?

**Comment by symphorien** [comment]
Just run `os-prober` as root. It should output one line by OS detected.

**Comment by florianjacob** [comment]
@symphorien Thanks! Now I was able to see the problem is detecting the other OS, not the grub.cfg generation. 👍

**Pull request merged**